Our.Umbraco.TagHelpers/Services/SelfHostService.cs

```diff
@@ -0,0 +1,132 @@
+﻿using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Our.Umbraco.TagHelpers.Extensions;
+using Our.Umbraco.TagHelpers.Models;
+using System;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Umbraco.Cms.Core.Cache;
+using Umbraco.Cms.Core.Logging;
+using Umbraco.Extensions;
+
+namespace Our.Umbraco.TagHelpers.Services
+{
+    public class SelfHostService : ISelfHostService
+    {
+        private readonly IProfilingLogger _logger;
+        private readonly IAppPolicyCache _runtimeCache;
+        private readonly IWebHostEnvironment _hostingEnvironment;
+        private readonly IConfiguration _config;
+
+        public SelfHostService(
+            IProfilingLogger logger,
+            IAppPolicyCache appPolicyCache,
+            IWebHostEnvironment hostingEnvironment,
+            IConfiguration config
+            )
+        {
+            _logger = logger;
+            _runtimeCache = appPolicyCache;
+            _hostingEnvironment = hostingEnvironment;
+            _config = config;
+        }
+
+        public async Task<SelfHostedFile> SelfHostFile(string url, string? subfolder = null, string? fileExtension = null)
+        {
+            return await _runtimeCache.GetCacheItem($"Our.Umbraco.TagHelpers.Services.SelfHostService.SelfHostedFile({url}, {subfolder}, {fileExtension})", async () =>
+            {
+                using (_logger.TraceDuration<ISelfHostService>($"Start generating SelfHostedFile: {url}", $"Finished generating SelfHostedFile: {url}"))
+                {
+                    var uri = new Uri(url, UriKind.Absolute);
+
+                    var selfHostedFile = new SelfHostedFile()
+                    {
+                        ExternalUrl = url,
+                        FileName = uri.Segments.Last() + fileExtension.IfNotNull(ext => ext.EnsureStartsWith(".")),
+                        FolderPath = GetFolderPath(uri, subfolder)
+                    };
+
+                    selfHostedFile.Url = await GetSelfHostedUrl(selfHostedFile);
+                    return selfHostedFile;
+                }
+            });
+        }
+
+        private string GetFolderPath(Uri uri, string? subfolder = null)
+        {
+            var folderPath = _config["Our.Umbraco.TagHelpers.SelfHost.RootFolder"].IfNullOrWhiteSpace("~/assets"); ;
```

Comment:
Was the intention for it to be this key @skttl ?
Can we document it in the readme and perhaps align/sync it with the image one Andy Boot done.

Reply:
Yes, but I've aligned with Andys now 👍